xen/drivers/passthrough/arm/scmi_smc_maker.c

```diff
@@ -0,0 +1,350 @@
+#define SCMI_NODE_COMPATIBLE   "arm,scmi-smc"
+#define SCMI_SHMEM_COMPATIBLE  "arm,scmi-shmem"
+
+struct scmi_phandle {
+    struct list_head list;
+    uint32_t phandle;
+    char full_name[SCMI_NODE_PATH_MAX_LEN];
+};
+
+LIST_HEAD(scmi_ph_list);
+
+static int __init mem_permit_access(struct domain *d, uint64_t addr, uint64_t len)
+{
+    int res;
+    res = iomem_permit_access(d, paddr_to_pfn(addr),
+            paddr_to_pfn(PAGE_ALIGN(addr + len - 1)));
+    if ( res )
+        return res;
+
+    return map_regions_p2mt(d, gaddr_to_gfn(addr), PFN_DOWN(len),
+            maddr_to_mfn(addr), p2m_mmio_direct_nc);
+}
+
+ int __init make_scmi_shmem_node(struct kernel_info *kinfo)
+{
+    int res;
+    void *fdt = kinfo->fdt;
+    char buf[64];
+    __be32 reg[GUEST_ROOT_ADDRESS_CELLS + GUEST_ROOT_SIZE_CELLS];
+    __be32 *cells;
+    struct domain *d = kinfo->d;
+
+#ifdef CONFIG_ARM_32
+    snprintf(buf, sizeof(buf), "scmi-shmem@%lx",
+            d->arch.sci->paddr);
+#else
+    snprintf(buf, sizeof(buf), "scmi-shmem@%lx",
+            d->arch.sci->paddr);
+#endif
+
+    if ( !is_domain_direct_mapped(d) )
+    {
+        printk(XENLOG_ERR "Non-direct mapped domains doesn't support SCMI\n");
+        return -ENODEV;
+    }
+
+    res = mem_permit_access(kinfo->d, kinfo->d->arch.sci->paddr, 0x1000);
+    if ( res )
+        return res;
+
+    res = fdt_begin_node(fdt, buf);
+    if ( res )
+        return res;
+
+    res = fdt_property_string(fdt, "compatible", SCMI_SHMEM_COMPATIBLE);
+    if ( res )
+        return res;
+
+    cells = &reg[0];
+    dt_child_set_range(&cells, GUEST_ROOT_ADDRESS_CELLS,
+                          GUEST_ROOT_SIZE_CELLS, d->arch.sci->paddr,
+                          GUEST_SCI_SHMEM_SIZE);
+
+    res = fdt_property(fdt, "reg", reg, sizeof(reg));
+    if ( res )
+        return res;
+
+    res = fdt_property_cell(fdt, "phandle", kinfo->phandle_sci);
+    if ( res )
+        return res;
+
+    res = fdt_end_node(fdt);
+    if ( res )
+        return res;
+
+    return 0;
+}
+
+/*
+ * The following methods are needed to get node name for the node full_path.
+ * This was done because some calls, such as dt_node_name return name
+ * before "@". So for node "protocol@19" it will return "protocol".
+ */
+static const char *dt_node_name_from_path(const struct dt_device_node *node)
+{
+    return strrchr(dt_node_full_name(node), '/') + 1;
+}
+
+static const char *name_from_path(const char *path)
+{
+    return strrchr(path, '/') + 1;
+}
+
+static int __init copy_properties(const struct dt_device_node *node, void* fdt)
+{
+    int rc;
+    const struct dt_property *pp;
+
+    printk(XENLOG_DEBUG "scmi_node_maker: copy properties for node: %s\n",
+           dt_node_name_from_path(node));
+
+    dt_for_each_property_node(node, pp)
+    {
+        /* Skipping phandle nodes in xen device-tree */
+        if ( dt_property_name_is_equal(pp, "phandle") ||
+             dt_property_name_is_equal(pp, "linux,phandle") )
```

Comment:
this is not enough unfortunately to be universal - the following props for arm,scmi.yaml has to be skipped:
-       mbox-names:
-      mboxes:
 -     shmem: 
 - linaro,optee-channel-id

May be it better to have consts array of props to be skipped

Reply:
This method related only for scmi child nodes. Scmi node itself is generating so no need to handle all these properties.